### PR TITLE
Evohome

### PR DESCRIPTION
**Pull request by foxy82**
It's starting to work with status updates!

Handler base class now we update state correctly for temperature. Set Temperature is being returned as 0.0 so need to find out why. Also we get NumberFormatException on some devices as the zoneId is coming out as
123.0 and the .0 bit is incorrect.

### CODE REVIEW

File: addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeHeatingZoneHandler.java [hunk 1]
```diff
@@ -14,30 +14,41 @@
...
     }
 
     @Override
     public void handleCommand(ChannelUID channelUID, Command command) {
+        if(command instanceof RefreshType){
+            if(getBridge() != null && getBridge().getHandler() != null){
+                EvohomeApiClient apiClient = ((EvohomeGatewayHandler) getBridge().getHandler()).getApiClient();
+                int zoneId = Integer.valueOf(getThing().getProperties().get(EvohomeBindingConstants.ZONE_ID));
+                int locationId = Integer.valueOf(getThing().getProperties().get(EvohomeBindingConstants.LOCATION_ID));
+                ZoneStatus zoneStatus = apiClient.getHeatingZone(locationId, zoneId);
```
Comment (Nebula83) [inline addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeHeatingZoneHandler.java]: I think you are safe to assume that a `zone id` is unique, so `locationId` isn't strictly necessary. That said, I does help the search routine.

File: addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeHeatingZoneHandler.java [hunk 2]
```diff
@@ -60,24 +71,11 @@ public void initialize() {
...
-                return;
-            }
-
-            updateState(EvohomeBindingConstants.SYSTEM_MODE_CHANNEL, new DecimalType(zoneStatus.temperature.temperature));
-            updateState(EvohomeBindingConstants.SET_POINT_CHANNEL, new DecimalType(zoneStatus.heatSetpoint.targetTemperature));
-            updateState(EvohomeBindingConstants.SET_POINT_STATUS_CHANNEL, new StringType(zoneStatus.heatSetpoint.setpointMode));
-            updateStatus(ThingStatus.ONLINE);
+            updateZoneStatus(zoneStatus);
         }
         // TODO: Initialize the thing. If done set status to ONLINE to indicate proper working.
```
Comment (Nebula83) [inline addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeHeatingZoneHandler.java]: Comments can go eventually

File: addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeHeatingZoneHandler.java [hunk 3]
```diff
@@ -86,4 +84,36 @@ public void initialize() {
...
     }
+
+    @Override
+    public void update(EvohomeApiClient client) {
+        String locationId = getThing().getProperties().get(EvohomeBindingConstants.LOCATION_ID);
+        String zoneId = getThing().getProperties().get(EvohomeBindingConstants.ZONE_ID);
+        logger.debug("Updating thing[{}] locationId[{}] zoneId[{}]", getThing().getLabel(), locationId, zoneId);
+        ZoneStatus zoneStatus = client.getHeatingZone(Integer.parseInt(locationId), Integer.parseInt(zoneId));
+        updateZoneStatus(zoneStatus);
+
```
Comment (Nebula83) [inline addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeHeatingZoneHandler.java]: whitespace (yes, I was brought up to nitpick in reviews ;-) )

**Comment by Nebula83** [comment]
Phat! It works! I get actual zone temps, super nice!
I did have to throw away all my evohome related things for the discovery to work again but I think that is normal.

**Comment by Nebula83** [comment]
By the way

> Handler base class now we update state correctly for temperature. Set Temperature is being returned as 0.0 so need to find out why. Also we get NumberFormatException on some devices as the zoneId is coming out as
123.0 and the .0 bit is incorrect.

Yeah, that's not right. The ApiClient logs all responses on the debug trace. Could you maybe privately send me the JSON repsonse data for tese items? 

I also found the issue with the zone target temp: in `HeatSetpointStatus` the `@SerializedName("TargetTemperature")` should be `@SerializedName("targetTemperature")`. If you change that it'll work =).

**Comment by foxy82** [comment]
It is very strange - I can't see what is casting it. I was thinking we could change the zoneId, loactionId etc to Strings in the model - after all they don't need to have arithmetic done on them...

**Comment by Nebula83** [comment]
Oh, yeah, might as well, smart =) Go ahead, no problems on my side. I think
it has to do with the internals of openhab, a bug causing that conversion
to bork?

2017-06-02 13:32 GMT+02:00 foxy82 <notifications@github.com>:

**Comment by foxy82** [comment]
Yeah my json comes in as "12345". Gson makes it an int and I was adding it straight to the properties. Then somewhere along the way I retrieve it and I get "12345.0" - it would also only happen on some zones. Very odd. I can't believe there could be such a fundamental bug in openhab but I can't see how I can be doing it!

**Pull request merged**